### PR TITLE
add recall @ k for retriever eval

### PR DESCRIPTION
**Pull request by bbqiu**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/bbqiu/mlflow/pull/10144?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/10144/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 10144
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

1. Add recall_at_k
2. remove "v1"
3. Add the k value in the metrics name like: precision_at_3
4. Refactor docs

### How is this PR tested?

- [ ] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/metrics/metric_definitions.py
Comment (liangz1) [inline mlflow/metrics/metric_definitions.py:380]: Vote for `len(ground truth)`

File: docs/source/python_api/mlflow.metrics.rst
Comment (liangz1) [inline docs/source/python_api/mlflow.metrics.rst:92]:
```suggestion
The following metrics are built-in metrics for the ``'retriever'`` model type, including precision_at_k and recall_at_k. They will be automatically calculated with a default ``k`` value of 3. 
```

File: docs/source/python_api/mlflow.metrics.rst
Comment (liangz1) [inline docs/source/python_api/mlflow.metrics.rst:115]:
```suggestion
To evaluate document retrieval models, use a dataset with the following columns: 
- Input queries
- Retrieved relevant doc IDs
- Ground-truth relevant doc IDs

Alternatively, you can also provide a function through the ``model`` parameter
to represent your retrieval model. The function should accept a Pandas 
DataFrame containing input queries and ground-truth relevant doc IDs, and return a DataFrame with a column of retrieved relevant doc IDs.

A "doc ID" is a string that uniquely identifies a document. Ensure that all doc IDs are 
represented as tuples of doc ID strings.

Parameters:
- ``targets``: Specifies the column name for ground-truth relevant doc IDs.
- ``predictions``: If using a static dataset: Specifies the column name for the retrieved relevant doc IDs.
- ``k``: A positive integer indicating how many retrieved doc IDs to consider for each input 
  query. Defaults to 3. You can change this value using the :py:func:`mlflow.evaluate` API in two 
  ways:
    1. Set ``evaluator_config={"k": 5}``
    2. Use ``extra_metrics`` like:
        ```
        extra_metrics = [
            mlflow.metrics.precision_at_k(k=5),
            mlflow.metrics.precision_at_k(k=6), 
            mlflow.metrics.recall_at_k(4), 
            mlflow.metrics.recall_at_k(5)
        ]
        ```
    In the second method, the ``k`` value in ``evaluator_config`` is ignored. It's advised to 
    also omit the ``model_type``; otherwise, precision@3 will be computed in addition to other 
    metrics like precision@5, precision@6, recall@4, and recall@5.
```

File: mlflow/metrics/metric_definitions.py
Comment (liangz1) [inline mlflow/metrics/metric_definitions.py:371]:
```suggestion
            predictions, "precision_at_k", predictions_col_specifier
```

File: mlflow/metrics/metric_definitions.py
Comment (liangz1) [inline mlflow/metrics/metric_definitions.py:372]:
```suggestion
        ) or not _validate_and_fix_text_tuple_data(targets, "precision_at_k", targets_col_specifier):
```

**Review by liangz1 — APPROVED** [review]
Looks good overall!

File: docs/source/python_api/mlflow.metrics.rst
Comment (prithvikannan) [inline docs/source/python_api/mlflow.metrics.rst:89]: we are moving the LLM autoeval metrics to a `mlflow.metrics.genai` namespace in https://github.com/mlflow/mlflow/pull/10138 to indicate that these are different. should we do the same for retriever metrics? else this `mlflow.metrics` module will get very large. i dont think we have to block on this btw.

File: tests/evaluate/test_default_evaluator.py
Comment (prithvikannan) [inline tests/evaluate/test_default_evaluator.py:3522]: shall we add this in this PR :) ?

**Review by prithvikannan — COMMENTED** [review]
Overall LGTM, left a couple small comments. @bbqiu do we already have an example checked in for retriever eval? doesn't need to be super fancy -- something in `examples/evaluation/` that can run manually would be great.

File: tests/evaluate/test_default_evaluator.py
Comment (bbqiu) [reply tests/evaluate/test_default_evaluator.py:3522]: just getting things ready for bug bash -- tests will be added before merging haha

File: docs/source/python_api/mlflow.metrics.rst
Comment (bbqiu) [reply docs/source/python_api/mlflow.metrics.rst:95]: not blocking, but made a jira issue here for thinking abt [this here](https://databricks.atlassian.net/browse/ML-36139)

**Comment by bbqiu** [comment]
@mlflow-automation autoformat

**Comment by harupy** [comment]
Can you rebase this PR on master to run the lint check? We accidentally disabled it in https://github.com/mlflow/mlflow/pull/10179.

File: docs/source/python_api/mlflow.metrics.rst
Comment (liangz1) [inline docs/source/python_api/mlflow.metrics.rst:108]:
```suggestion
Alternatively, you can also provide a function through the ``model`` parameter to represent 
```

**Review by liangz1 — APPROVED** [review]
LGTM! Awesome work, @bbqiu !

File: mlflow/metrics/metric_definitions.py
Comment (harupy) [inline mlflow/metrics/metric_definitions.py:373]: /ai rewrite this using zip (assuming predictions and targets are the same size). No explanation is required. Just give me code.

File: docs/source/python_api/mlflow.metrics.rst
Comment (harupy) [inline docs/source/python_api/mlflow.metrics.rst:113]:
```suggestion
list of doc ID strings.
```

File: .github/workflows/master.yml
Comment (harupy) [inline .github/workflows/master.yml:35]: It looks like this change is related. Rebase might be required.

File: docs/source/python_api/mlflow.metrics.rst
Comment (harupy) [inline docs/source/python_api/mlflow.metrics.rst:132]: Suprised that this renders fine lol

File: docs/source/python_api/mlflow.metrics.rst
Comment (harupy) [inline docs/source/python_api/mlflow.metrics.rst:124]: I'd use this for clarity:

```python
mlflow.evaluate(
    ...,
    evaluator_config={"retriever_k": 5}
)
```

File: docs/source/python_api/mlflow.metrics.rst
Comment (bbqiu) [reply docs/source/python_api/mlflow.metrics.rst:132]: i was too, but I think this is the best way to handle multil-line python code in our docs

File: mlflow/metrics/metric_definitions.py
Comment (bbqiu) [reply mlflow/metrics/metric_definitions.py]: predictions and targets are the same size, so this works! thanks for the recommendation

File: docs/source/python_api/mlflow.metrics.rst
Comment (dbczumar) [inline docs/source/python_api/mlflow.metrics.rst:114]: `All doc IDs should be entered as a list of doc ID strings.` is a bit confusing. Can we clarify?

File: docs/source/python_api/mlflow.metrics.rst
Comment (dbczumar) [inline docs/source/python_api/mlflow.metrics.rst:133]: Can we update these code blocks to include example `targets` and `predictions`?

File: mlflow/models/evaluation/base.py
Comment (dbczumar) [inline mlflow/models/evaluation/base.py:419]: instead of `json.dumps(val)`, can we do `str(val)`? This will handle the case where the data isn't JSON-serializable, e.g. you're dealing with a list of objects

File: mlflow/models/evaluation/default_evaluator.py
Comment (dbczumar) [inline mlflow/models/evaluation/default_evaluator.py:1717]: This validation helps for `model_type="retriever", but I tested with `extra_metrics` and can still specify k=0, k=-2, etc. For example:

```
import pandas as pd
import mlflow

X = pd.DataFrame({"question": ["q1?"] * 3, "ground_truth": [["doc1", "doc2"]] * 3})

def fn(X):
    return {"retrieved_context": [["doc1", "doc3", "doc2"]] * len(X)}

with mlflow.start_run() as run:
    results = mlflow.evaluate(
        model=fn,
        data=X,
        targets="ground_truth",
        extra_metrics=[mlflow.metrics.precision_at_k(-1), mlflow.metrics.recall_at_k(-2)],
    )
run = mlflow.get_run(run.info.run_id)
print(run.data.metrics)
```

yields

```
{'precision_at_-1/variance': 0.0, 'precision_at_-1/mean': 0.5, 'precision_at_-1/p90': 0.5, 'recall_at_-2/variance': 0.0, 'recall_at_-2/mean': 0.5, 'recall_at_-2/p90': 0.5}

```

Comment (dbczumar) [reply mlflow/models/evaluation/default_evaluator.py:1717]: Can we add validation to ensure that constructing `mlflow.metrics.precision_at_k(K<=0 || k non-integer)` throws?

Comment (bbqiu) [reply mlflow/models/evaluation/default_evaluator.py]: yep!

**Review by dbczumar — APPROVED** [review]
LGTM!

**Pull request merged**
